### DISCUSSION TITLE
Tuning prognostic momentum fluxes

### DISCUSSION THREAD

**Discussion by bstephens82** [NCAR/amwg_dev#216] [TF2: Tuning]
As a second track for tuning work, we are going to try to improve the results of turning on CLUBB's prognostic momentum fluxes for CAM7.  We have a somewhat decent looking result already (https://webext.cgd.ucar.edu/F2000climo/d92pmf/atm/, this is just the out-of-the-box 092 tag with cam_dev, L58 and prognostic momentum fluxes turned on), but we are hoping to be able to improve this.  The main parameter to tune is C_uu_shr, so I will start a series of runs with a range of values for this parameter and we can see if we find an improved result. 

@vlarson @JulioTBacmeister @adamrher @zarzycki

**Comment by vlarson** [comment]
Looking at the prognostic momentum flux equation, boosting `C_uu_shr` should decrease the effective diffusivity for momentum.  I.e., `C_uu_shr` should offset the turbulent production term, `wp2*dum/dz`.  I have no theory for what the range of `C_uu_shr` should be, so I'd recommend varying it widely, between, say, 0.0 and 0.9.

**Reply by bstephens82** [reply]
Okay, I have these runs in the queue.

**Comment by bstephens82** [comment]
Thomas noticed there was a one-line bug in CLUBB's mono_flux_limiter.F90 code, and we've been thinking about what to do with this.  When running without prognostic momentum fluxes, the bug is not relevant because the MFL does not operate on um and vm.  But when running with PMF, it might make a difference.  To properly fix the bug we would need to make a change to the CLUBB repo and then update the external in CAM.  Another idea was to see if CAM would run stably with prognostic momentum fluxes but with the MFL turned off for um and vm.  Thanks to the 091 CLUBB external updates, this is easy with namelist changes.  Here are the results of a one-year test I did, with the 099 tag, using PMF but with MFL turned off for um and vm.  I am continuing this test for 5 years and will update with those diagnostics when I have them.

https://webext.cgd.ucar.edu/F2000climo/f2000.099.pmf.cuushr01.c701_nomflumvm/atm/f2000.099.pmf.cuushr01.c701_nomflumvm-obs.0001_1/

**Reply by zarzycki** [reply]
Ben, thanks for doing this. I am curious what the 5-year diagnostics will look like since I don't trust 1-year diagnostics for some of the dynamical quantities we are interested in.

Can you also grep through the log file to see how often the MFL is invoked? I do recall CLUBB logs a statement each time that code is breached, but can u/v be disentangled from the MFL being triggered for other variables? If not, perhaps just saying "MFL triggered X times in PMF/MFLon and Y times in PMF/MFLoff" is enough for government work?

**Reply by bstephens82** [reply]
Well in this case the MFL is turned off for um and vm so it will never trigger for those, I could take a look at some other 099 runs I have to see if they give some indication of what happens when it's active for the winds.  If you're interested in how often it was triggered for thlm and rtm I can check the log files.

**Comment by bstephens82** [comment]
Unfortunately my 099 tag L58 cam_dev prognostic momentum flux run **without** the monotonic flux limiter for um and vm (but still active for thlm and rtm) crashed with an error from CLUBB in the 6th year, seemingly in August since the July mean file was printed out.  The error comes from near 30 lat and 85 long, which I think is around the Himalayas which has given us trouble with the MFL before.  Anyway NaNs show up in various fields and then we get a crash.  I don't have high-frequency output from this run so I'm not sure I would be able to pinpoint the specific behavior that led to the crash...

By contrast, my 092 tag L58 cam_dev runs with PMF and MFL active (and including Thomas's bug fix) are still going after 15+ years.  So, barring some other cause from changes coming from the 093-099 tags, it seems turning off the MFL for um and vm does cause stability issues.  If so, it may be best to fix the bug (which just FYI is already fixed in UWM's CLUBB repo) rather than to turn off MFL. 

@JulioTBacmeister @vlarson @adamrher @Katetc @zarzycki any thoughts?

**Comment by JulioTBacmeister** [comment]
That's frustrating, but for the next round of coupled testing I think we should just proceed with the bugfix.  I do think it would be worth exploring this further in the future,  and let's have a look at the climate in this run. 5+ years should be enough to show if removing the MFL from prognostic momentum flux makes a big difference.

**Reply by adamrher** [reply]
It is what it is. As we all will want to be running with Thomas' bug fix, I propose three steps:

1. add the bug fix to our CLUBB externals
2. open an issue to bring in new CLUBB externals
3. bring in new CLUBB externals

Where we can just take our time with (3). (1) is preferable to SourceMod'ing in the change every time we want to run the workhorse. I don't see why we can't just make this one exception and modify our externals for this bug fix.

**Reply by bstephens82** [reply]
How would we add one change to the CLUBB external?  Wouldn't that take place on the UWM side?  If so, would we take the existing CLUBB external tag, branch from it, make the change, and then give it the same tag name?

**Reply by zarzycki** [reply]
I am interested in the climate -- hopefully that looks the same?

Again, if the MFL is being called relatively infrequently (like I think it is based on my logfiles), I am fine with applying the bugfix to prevent these edge cases from being a problem stability-wise. I suspect the higher vertical resolution is more stressful w/ the relatively "long" timesteps we take, but in some ways that's what an MFL aims to help with.

I will defer to everyone else as to the best way to update it ... I agree with @adamrher that it would be nice not to bring in a whole new CLUBB external from UWM just to patch a few lines...

**Reply by bstephens82** [reply]
I'll have the 5-year diagnostics in a few minutes and upload the link here.

**Reply by adamrher** [reply]
@zarzycki I did look at the time-step sensitivity to the rtm spikes, and it was marginal at best ([see table in this post](https://github.com/NCAR/amwg_dev/discussions/134#discussion-4154451)). That's a different beast than the um/vm spikes, but I think it's useful in that it suggests the mfl activity doesn't significantly change when you halve the clubb time-step.

@bstephens82 oh I forgot we no longer store our externals in https://github.com/ESCOMP/CLUBB_CESM, and instead they are hosted at UWM. Not sure what the easiest way to run with bug fix is then ...

**Reply by zarzycki** [reply]
BTW, a bug in the ear (perhaps this is for @vlarson and UWM) ... currently the MFL warning in the log file looks like...

```
226: Vertical integral of xm is zero;mfl will remove spike at top of domain,
226: but it will not conserve xm.
```

maybe I'm missing something. However, it would be good if we could see exactly what state variable (i.e., um, vm, etc.) are triggering each call instead of the generic "xm" ... then we'd have some diagnostics as to how often the MFL is being called and for what variables (and time, lat, lon could also be helpful :))

**Comment by tto061** [comment]
Hi, the master in clubb_release has the simple bug fix in, perhaps this 
could be merged into the current tag?

and yes adding solve_type to the printout there might be helpful, but
I still wonder if, when solve_type=mono_flux_u/vm (i.e. 4 or 5), perhaps 
a simple additive unconditional fix, skipping that magnitude test 
entirely, would not be more appropriate.

Thomas Toniazzo
MISU          Stockholms university
SE-106 91 Stockholm         Sverige

**Comment by bstephens82** [comment]
The 5-year diagnostics for the run with prognostic momentum flux but **without** monotonic flux limiter are here:
https://webext.cgd.ucar.edu/F2000climo/f2000.099.pmf.cuushr01.c701_nomflumvm/atm/f2000.099.pmf.cuushr01.c701_nomflumvm-obs.0001_5/

Just as a reminder, this run crashed in the 6th year.

**Comment by bstephens82** [comment]
Here are the model vs. model diagnostics for the 5-year results with and without the MFL active for um and vm (last post above has the model vs. obs):
https://webext.cgd.ucar.edu/F2000climo/f2000.099.pmf.cuushr01.c701_nomflumvm/atm/f2000.099.pmf.cuushr01.c701_nomflumvm.0001_5-f2000.099.pmf.cuushr01.c701.0001_5/

Also bear in mind these are using the 099 tag which has a higher-magnitude SWCF (about -50 W/m2).

**Reply by adamrher** [reply]
In the wind stress, [these differences](https://webext.cgd.ucar.edu/F2000climo/f2000.099.pmf.cuushr01.c701_nomflumvm/atm/f2000.099.pmf.cuushr01.c701_nomflumvm.0001_5-f2000.099.pmf.cuushr01.c701.0001_5/set5_6/set6_ANN_STRESS_MAG_c.png) look like the differences of the pair of runs you did with MFL on, but w/ and w/o the bug fix (you showed this at last Friday's mtg). The bug fix causes positive (red) wind stress changes, especially in the Southern Ocean, whereas these latest results are showing that the wind stress biases are negative (blue) when you turn on the MFL limiter (and I'm assuming MFL on here also has the bug fix). The change in wind stress due to fixing the bug in the MFL limiter is similar to the change of just turning on the MFL limiter w/ the bug fix. This suggests that the change in wind stress of turning on the MFL limiter **with its bug intact** would be comparatively small.

**Comment by vlarson** [comment]
I am passing along a comment from Dmitrii Mironov about the "barotropic" nature of the wind bias over the Southern Ocean.  In other words, compared to MERRA, CAM's zonal wind over the Southern Ocean (60 S) is [too strong at _all altitudes_](https://webext.cgd.ucar.edu/F2000climo/f2000.099.pmf.cuushr01.c701_nomflumvm/atm/f2000.099.pmf.cuushr01.c701_nomflumvm-obs.0001_5/set4/set4_ANN_U_MERRA_obsc.png).  

He points out that a column-based turbulence parameterization can only _directly_ re-arrange the momentum (or heat or moisture) in the vertical.  (Indirectly it can have other effects as well, but let's ignore those for the moment.)  The vertical integral of the wind is only affected by surface friction, which drains the column of momentum.  

This observation hints that the excessive wind might be related to thermal wind balance.  Indeed, CAM is much [colder than MERRA over the South Pole stratosphere](https://webext.cgd.ucar.edu/F2000climo/f2000.099.pmf.cuushr01.c701_nomflumvm/atm/f2000.099.pmf.cuushr01.c701_nomflumvm-obs.0001_5/set4/set4_ANN_T_MERRA_obsc.png).  Does anyone happen to know what causes this cold bias, how to fix it, or why CLUBB might be indirectly affecting it?

**Reply by zarzycki** [reply]
Is the hypothesis that removing the cold stratospheric bias would improve the surface stress?

Would an easy test of this hypothesis be to nudge the upper atmosphere temperature to reanalysis (say < 200mb) to remove the cold bias and observe whether the surface stress improves?

**Reply by vlarson** [reply]
> Is the hypothesis that removing the cold stratospheric bias would improve the surface stress?

Yes, the hypothesis is that the shear in the Southern Ocean jet is linked, by thermal wind balance, to the gradient in temperature from the stratospheric South Pole to the midlatitudes.  If the temperature gradient is too strong, the jet shear will be too strong.  I'm not sure how the shear aloft would influence the surface, but the hypothesis is that it does have an effect.

**Reply by adamrher** [reply]
Our new candidate compset is significantly warmer in the 40km-20km range. See https://github.com/NCAR/amwg_dev/issues/229#issuecomment-1493005782. It does [reduce the strength of the westerly jets](https://webext.cgd.ucar.edu/FLTHIST/f.cam6_3_101.FLTHIST_v0a.ne30.001/atm/f.cam6_3_101.FLTHIST_v0a.ne30.001_1979_1985_vs_f.cam6_3_100.FWscHIST.ne30_L58.001_1979_1985/html_img/plot_page_U_ANN_Zonal_Mean.html), although the changes are not strictly barotropic, and it looks like the SH jet is less impacted. Note also this is only a 5 year average. 

Separately, I do know that you can impact T near the model top by increasing the sponge layer diffusion coef. in the se-dycore (se_nu_top) ... since this will increase frictional heating.

I don't think either of these approaches will attack the cold bias where it is largest, in the lower stratosphere (about 12km according to the MERRA bias plot Vince linked to). But I suppose they could chip away at it. I'd be curious to hear what Julio or others know about this T bias. My understanding is that it is a well-known, long standing problem. Unfortunately, that's the extent of my understanding.

**Comment by islasimpson** [comment]
We have some runs with CESM2 FV, low top, with the stratospheric zonal mean u and v nudged toward MERRA2, which may be useful since T probably adjusts to balance it.  
/glade/campaign/cgd/cas/islas/QBOMJO_NUDGE2MERRA
